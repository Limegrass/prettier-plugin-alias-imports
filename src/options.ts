import { ParserOptions, SupportOption } from "prettier";

export const SupportedParsers = {
    TypeScript: "typescript",
    Babel: "babel",
    Flow: "flow",
} as const;

/** Type for the values of `SupportedParsers`. */
export type SupportedParserName =
    typeof SupportedParsers[keyof typeof SupportedParsers];

export interface AliasImportOptions extends ParserOptions {
    aliasConfigPath?: string;
    parser: SupportedParserName;
}

/**
 * Force `options` to only include keys from the AliasImportOptions interface
 */
type AliasImportSupportOptions = Omit<
    Record<keyof AliasImportOptions, SupportOption>,
    keyof ParserOptions
>;

/**
 * Configuration options specific to this plugin
 * which prettier presents to the user if installed.
 */
export const options: AliasImportSupportOptions = {
    // Would be nice to be able to use some type
    // type ConfigFile = { path: string; type: SupportedParserPlugins }
    // but I don't think that is supported
    aliasConfigPath: {
        since: "0.1.0",
        type: "path", // string is supported by prettier but the types needs to be updated
        category: "Global",
        description: "Configuration file to read. Defaults to tsconfig.json",
    },
};
