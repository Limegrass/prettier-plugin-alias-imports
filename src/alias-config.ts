import { log } from "#src/log";
import { AliasImportOptions } from "#src/options";
// Unsure how to best support find-up@^6.0.0 with the change to modules
// May depend on https://github.com/prettier/prettier/issues/8772
import { sync as findUpSync } from "find-up";
import micromatch from "micromatch";
import { dirname, join as joinPath } from "path";
import {
    ConfigLoaderResult,
    ConfigLoaderSuccessResult,
    loadConfig,
} from "tsconfig-paths";

/**
 * Failed loads result from invalid tsconfigs or tsconfigs without baseUrl.
 * Casts input into a ConfigLoaderSuccessResult result if valid.
 */
const isConfigLoaderSuccessResult = (
    configLoaderResult: ConfigLoaderResult
): configLoaderResult is ConfigLoaderSuccessResult =>
    configLoaderResult.resultType === "success";

export const loadAliasConfigs = (
    options: AliasImportOptions
): AliasConfig[] => {
    const configFilePath = findUpSync(
        [options.aliasConfigPath, "tsconfig.json", "jsconfig.json"].filter(
            Boolean
        ) as string[],
        {
            cwd: options.filepath, // works as dir
        }
    );

    if (!configFilePath) {
        log("cannot find ts/jsconfig, try assigning aliasConfigPath");
        return []; // Not sure if plugins should throw errors
    }

    const config = loadConfig(configFilePath);

    if (!isConfigLoaderSuccessResult(config)) {
        log("set compilerOptions.baseUrl and validate ts/jsconfig provided");
        return []; // Not sure if plugins should throw errors
    }

    const configDir = dirname(configFilePath);
    const absoluteRootDir = joinPath(configDir, config.baseUrl);

    return Object.entries(config.paths).reduce(
        (configs, [aliasGlob, aliasPaths]) => {
            aliasPaths.forEach((aliasPath) => {
                const relativePathBase = micromatch.scan(aliasPath).base;
                configs.push({
                    alias: micromatch.scan(aliasGlob).base,
                    path: {
                        absolute: joinPath(absoluteRootDir, relativePathBase),
                    },
                });
            });
            return configs;
        },
        [] as AliasConfig[]
    );
};

export interface AliasConfig {
    alias: string;
    path: {
        /** Glob values using absolute paths */
        absolute: string;
    };
}
