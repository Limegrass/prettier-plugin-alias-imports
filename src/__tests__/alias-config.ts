import { loadAliasConfigs } from "#src/alias-config";
import { AliasImportOptions, SupportedParsers } from "#src/options";
import { DEFAULT_OPTIONS } from "#src/__tests__/__mocks__/options";
import { sync as findUpSync } from "find-up";
import { loadConfig } from "tsconfig-paths";

jest.mock("tsconfig-paths");
jest.mock("find-up", () => ({
    sync: jest.fn(),
}));

const CURRENT_FILE_DIR = "src";
const TEST_OPTIONS = Object.freeze({
    ...DEFAULT_OPTIONS,
    filepath: `${CURRENT_FILE_DIR}/test.ts`,
    plugins: [],
});

describe.each([
    { parser: SupportedParsers.TypeScript },
    { parser: SupportedParsers.Babel },
])("loadAliases($parser)", ({ parser }) => {
    const options: AliasImportOptions = {
        ...TEST_OPTIONS,
        parser,
    };

    // Default happy path
    beforeEach(() => {
        const alias = "test";
        const pattern = "potato/*";
        const paths = {
            [alias]: [pattern],
        };
        (loadConfig as jest.Mock).mockReturnValue({
            resultType: "success",
            baseUrl: "/git/proj/tsconfig.json",
            paths,
        });
        (findUpSync as unknown as jest.Mock).mockReturnValue("tsconfig.json");
    });

    it("returns mapped alias and patterns for valid tsconfig", () => {
        const configs = loadAliasConfigs(options);
        expect(configs).toMatchSnapshot();
    });

    it("uses aliasConfigPath from options", () => {
        const customConfigPath = "potato.com";
        options.aliasConfigPath = customConfigPath;
        loadAliasConfigs(options);
        expect(findUpSync).toHaveBeenCalledWith(
            expect.arrayContaining([options.aliasConfigPath])
        );
    });

    it("returns an empty array if config path cannot be located", () => {
        const customConfigPath = "potato.com";
        options.aliasConfigPath = customConfigPath;
        (findUpSync as unknown as jest.Mock).mockReturnValue(undefined);
        const configs = loadAliasConfigs(options);
        expect(findUpSync).toMatchSnapshot();
        expect(configs).toHaveLength(0);
    });

    it("returns an empty array if the tsconfig has errors", () => {
        (loadConfig as jest.Mock).mockReturnValue({
            resultType: "failed",
        });
        const configs = loadAliasConfigs(options);
        expect(findUpSync).toMatchSnapshot();
        expect(configs).toHaveLength(0);
    });
});
