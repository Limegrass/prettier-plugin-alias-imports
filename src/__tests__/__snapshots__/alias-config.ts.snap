// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadAliases(babel) returns mapped alias and patterns for valid tsconfig 1`] = `
Array [
  Object {
    "alias": "test",
    "path": Object {
      "absolute": "git/proj/tsconfig.json/potato",
    },
  },
]
`;

exports[`loadAliases(typescript) returns mapped alias and patterns for valid tsconfig 1`] = `
Array [
  Object {
    "alias": "test",
    "path": Object {
      "absolute": "git/proj/tsconfig.json/potato",
    },
  },
]
`;
